### PR TITLE
Enhance CPE with Add Fragment to Ext Point extension

### PR DESCRIPTION
**Pull request by nikmace**
Feat for #1360.

- Includes changes to the Control Property Editor outline tree. 
- Adds new dialog for preview-middleware-client.
- Enables creation of XML fragments for Extension Points from the outline tree (when right-clicking on extension point) or from the application (when clicking on control).

### CODE REVIEW

File: packages/preview-middleware-client/src/adp/controllers/BaseDialog.controller.ts [hunk 1]
```diff
@@ -41,6 +43,53 @@ export default abstract class BaseDialog extends Controller {
...
+                this.dialog.getBeginButton().setEnabled(false);
+            } else {
+                this.dialog.getBeginButton().setEnabled(true);
+                source.setValueState(ValueState.None);
+                this.model.setProperty('/newFragmentName', fragmentName);
+            }
+        }
+    }
+
+    setEscapeHandler() {
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/adp/controllers/BaseDialog.controller.ts:86]: Please add JSDoc

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]
Please add changeset

File: packages/preview-middleware-client/src/adp/controllers/BaseDialog.controller.ts (hunk 1 above)
Comment (nikmace) [reply packages/preview-middleware-client/src/adp/controllers/BaseDialog.controller.ts:89]: Added in: https://github.com/SAP/open-ux-tools/pull/1378/commits/38e7284c4c05f224d4b0a45bc40f8e6b93ddffe2

Comment (zdravko-georgiev) [reply packages/preview-middleware-client/src/adp/controllers/BaseDialog.controller.ts:89]: thanks

**Comment by zdravko-georgiev** [comment]
@nikmace Please check the code smell

**Comment by zdravko-georgiev** [comment]
@nikmace I am trying to test locally, but no matter where I right-click in the Outline I do not see any special actions. What am I missing?

**Comment by zdravko-georgiev** [comment]
> > @nikmace I am trying to test locally, but no matter where I right-click in the Outline I do not see any special actions. What am I missing?
> 
> Depends with which application you are testing and if it has extension points.
> 
> The button to open the dialog is only visible when you right click on an extension point in the outline tree (extension points have an icon on the left and the text, when hovered, shows a tooltip which is a title of that element. the text of the tooltip is the full name of the extension point. this was done because the outline tree has no overflow on x-axis and you cannot scroll horizontally). NOTE: You can right click only on extension points, other nodes have no title and no tooltip when right clicking them, the default behavior of the context menu in the browser is preserved there.
> 
> Or you could open the same dialog when clicking on control inside the application that has an extension point, for example sap.m.Table (you would see a "add xml fragment at extension point" appear in the rta context menu).

Thanks. Apparently my app does not have extensions points. Do you have an application for me?

**Comment by nikmace** [comment]
> > > @nikmace I am trying to test locally, but no matter where I right-click in the Outline I do not see any special actions. What am I missing?
> > 
> > 
> > Depends with which application you are testing and if it has extension points.
> > The button to open the dialog is only visible when you right click on an extension point in the outline tree (extension points have an icon on the left and the text, when hovered, shows a tooltip which is a title of that element. the text of the tooltip is the full name of the extension point. this was done because the outline tree has no overflow on x-axis and you cannot scroll horizontally). NOTE: You can right click only on extension points, other nodes have no title and no tooltip when right clicking them, the default behavior of the context menu in the browser is preserved there.
> > Or you could open the same dialog when clicking on control inside the application that has an extension point, for example sap.m.Table (you would see a "add xml fragment at extension point" appear in the rta context menu).
> 
> Thanks. Apparently my app does not have extensions points. Do you have an application for me?

This one should work: sap.ui.demoapps.rta.fiorielements. If you have the repository [open-ux-tools-playground](https://github.tools.sap/ux/open-ux-tools-playground/), you can find it under apps/adp.v2app.

**Comment by zdravko-georgiev** [comment]
> > > > @nikmace I am trying to test locally, but no matter where I right-click in the Outline I do not see any special actions. What am I missing?
> > > 
> > > 
> > > Depends with which application you are testing and if it has extension points.
> > > The button to open the dialog is only visible when you right click on an extension point in the outline tree (extension points have an icon on the left and the text, when hovered, shows a tooltip which is a title of that element. the text of the tooltip is the full name of the extension point. this was done because the outline tree has no overflow on x-axis and you cannot scroll horizontally). NOTE: You can right click only on extension points, other nodes have no title and no tooltip when right clicking them, the default behavior of the context menu in the browser is preserved there.
> > > Or you could open the same dialog when clicking on control inside the application that has an extension point, for example sap.m.Table (you would see a "add xml fragment at extension point" appear in the rta context menu).
> > 
> > 
> > Thanks. Apparently my app does not have extensions points. Do you have an application for me?
> 
> This one should work: sap.ui.demoapps.rta.fiorielements. If you have the repository [open-ux-tools-playground](https://github.tools.sap/ux/open-ux-tools-playground/), you can find it under apps/adp.v2app.

I do have the repository, but this application also do not work 😕

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]
Please add also a changeset for the `preview-middleware` which will consume this changes.

File: packages/preview-middleware-client/src/adp/utils.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ */
+export function createDeferred<T>(): Deferred<T> {
+    let resolve: Deferred<T>['resolve'];
+    let reject: Deferred<T>['reject'];
+    const promise = new Promise<T>((res, rej) => {
+        resolve = res;
+        reject = rej;
+    });
+    return { promise, resolve: resolve!, reject: reject! };
+}
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/src/adp/utils.ts:22]: Is there a way to rewrite this so that we can remove the usage of the `!`.

Comment (zdravko-georgiev) [reply packages/preview-middleware-client/src/adp/utils.ts]: Thanks!

**Review by zdravko-georgiev — DISMISSED** [review]
Code changes look good
Changeset is OK
Review comments addressed
Manual test ✅

**Review by testojs — DISMISSED** [review]
- changes look good
- good test coverage
- main scenario is tested and working as expected

**Comment by nikmace** [comment]
@testojs @zdravko-georgiev There was some problem when I updated the PR with the changes from the main branch. SonarQube for some reason detected accessibility bugs that were not detected before the reviews.. The last commits do not change any functionality.

**Review by zdravko-georgiev — APPROVED** [review]

**Review by zdravko-georgiev — DISMISSED** [review]
Code changes look good
Changeset is OK
Review comments addressed
Very good test coverage
Did NOT tested manually

**Review by testojs — DISMISSED** [review]
* code improved with the latest changes
* manually tested - works as expected

**Comment by nikmace** [comment]
Would someone from @SAP/ux-tools-control-property-editor be able to review the code? Thanks in advance!

File: packages/preview-middleware-client/test/unit/cpe/outline/nodes.test.ts [hunk 3]
```diff
@@ -9,37 +12,33 @@ jest.mock('../../../../src/cpe/outline/utils', () => {
...
-            .fn()
-            .mockReturnValueOnce('Component')
-            .mockReturnValueOnce('Component')
-            .mockReturnValue('')
+        getProperty: jest.fn().mockReturnValueOnce('Component').mockReturnValueOnce('Component').mockReturnValue('')
     });
     describe('transformNodes', () => {
         test('empty tree', async () => {
-            expect(await transformNodes([])).toStrictEqual([]);
+            expect(await transformNodes([], 'UI_ADAPTATION')).toStrictEqual([]);
```
Comment (zdravko-georgiev) [inline packages/preview-middleware-client/test/unit/cpe/outline/nodes.test.ts:25]: Shouldn't this be `Adaptation_Project`?

**Review by zdravko-georgiev — CHANGES_REQUESTED** [review]

File: packages/preview-middleware-client/test/unit/cpe/outline/nodes.test.ts (hunk 3 above)
Comment (nikmace) [reply packages/preview-middleware-client/test/unit/cpe/outline/nodes.test.ts:25]: The test for the `Adaptation_Project` scenario is below with the name `extension point`. The one you are asking about is testing without the extension points enabled, meaning everything other than **Adaptation Project**.

Comment (zdravko-georgiev) [reply packages/preview-middleware-client/test/unit/cpe/outline/nodes.test.ts:25]: Got it, thanks!

**Review by zdravko-georgiev — APPROVED** [review]
Code changes look good
Changeset is OK
Review comments addressed
Very good test coverage
Did NOT tested manually

**Review by IvoSG — APPROVED** [review]
Code changes look good
Changeset is OK
Very good test coverage
Tested locally

**Pull request merged**